aiida_optimize/_optimization_workchain.py

```diff
@@ -138,12 +143,15 @@ def get_results(self):
             opt.update(outputs)
 
     @check_workchain_step
-    def finalize(self):
+    def finalize(self): # pylint: disable=inconsistent-return-statements
         """
         Return the output after the optimization procedure has finished.
         """
         self.report('Finalizing optimization procedure.')
         with self.optimizer() as opt:
+            self.report('Checking if optimizer is finished ok.')
```

Comment:
That's exactly the reason I have it as it is. I think it's much safer to have the workchain not return an output if convergence is not reached because even the "best" result is still not converged and therefore at best useless[1] and at worst harmful if the caller doesn't check explicitly that the optimization finished ok (using the is_finished_ok or is_converged if we switch to that). In future, if we make the optimization robust to some amount of evaluation failures, some non-zero exit statuses will also constitute success which even further limits the ability of callers to differentiate non-converged exit statuses from converged exit statuses if not explicitly checking is_finished_ok or is_converged, so returning no output in the failure state will be even more necessary to clearly differentiate failure from non-convergence.

[1] I say "useless" loosely here because many of the calculations / work chain iterations that are performed can be reused via caching to re-run the optimization with more inputs at much lower cost.

For now, I think we leave it as-is because it's a more reliable way to ensure that the engine is used as intended and its outputs are not misinterpreted.

Reply:
Yep, absolutely agree - let's leave it this way.